src/v/storage/mvlog/versioned_log.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "storage/mvlog/versioned_log.h"
+
+#include "container/fragmented_vector.h"
+#include "model/offset_interval.h"
+#include "storage/mvlog/file.h"
+#include "storage/mvlog/logger.h"
+#include "storage/mvlog/readable_segment.h"
+#include "storage/mvlog/segment_appender.h"
+
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <chrono>
+
+using namespace std::literals::chrono_literals;
+
+namespace storage::experimental::mvlog {
+
+active_segment::active_segment(
+  std::unique_ptr<file> f, model::offset o, segment_id id, size_t target_size)
+  : segment_file(std::move(f))
+  , appender(std::make_unique<segment_appender>(segment_file.get()))
+  , readable_seg(std::make_unique<readable_segment>(segment_file.get()))
+  , construct_time(ss::lowres_clock::now())
+  , target_max_size(target_size)
+  , id(id)
+  , base_offset(o)
+  , next_offset(o) {}
+
+active_segment::~active_segment() = default;
+
+readonly_segment::~readonly_segment() = default;
+readonly_segment::readonly_segment(std::unique_ptr<active_segment> active_seg)
+  : segment_file(std::move(active_seg->segment_file))
+  , readable_seg(std::move(active_seg->readable_seg))
+  , id(active_seg->id)
+  , offsets(model::bounded_offset_interval::checked(
+      active_seg->base_offset, model::prev_offset(active_seg->next_offset))) {}
+
+versioned_log::versioned_log(storage::ntp_config cfg)
+  : ntp_cfg_(std::move(cfg)) {}
+
+ss::future<> versioned_log::roll_unlocked() {
+    vassert(
+      !active_segment_lock_.ready(),
+      "roll_unlocked() must be called with active segment lock held");
+    vassert(active_seg_ != nullptr, "Expected an active segment");
+    vlog(
+      log.info,
+      "Rolling segment file {}",
+      active_seg_->segment_file->filepath().c_str());
+    co_await active_seg_->segment_file->flush();
```

Comment:
i wonder if the flush is necessary? if we extend the log and keep writing to the next segment at some point the "log" will be flushed, at which point we could flush all unflushed segments. is there a particular reason for flushing here before rolling?

Reply:
Good point, I felt it seemed natural for applying segment.ms for a flush to occur because it indicates we have some data that's otherwise just sitting in memory. But for segment.bytes maybe it's more worth it to be lazy about flushing

Reply:
yeh. we can always measure it and see if it is worth keeping or removing.